src/controller/CHIPDeviceControllerFactory.h

```diff
@@ -0,0 +1,117 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      DeviceControllerFactory is a singleton utility class that manages the
+ *      runtime DeviceControllerSystemState and provides APIs to setup DeviceControllers
+ *      and DeviceCommissioners.
+ *
+ *      Together with the SystemState this class implicitly manages the lifecycle of the underlying
+ *      CHIP stack. It lazily initializes the CHIPStack when setting up Controllers if the SystemState
+ *      was previously shutdown.
+ */
+
+#pragma once
+
+#include <controller/CHIPDeviceController.h>
+#include <controller/CHIPDeviceControllerSystemState.h>
+
+namespace chip {
+
+namespace Controller {
+
+struct SetupParams
+{
+#if CHIP_DEVICE_CONFIG_ENABLE_MDNS
+    DeviceAddressUpdateDelegate * deviceAddressUpdateDelegate = nullptr;
+#endif
+    OperationalCredentialsDelegate * operationalCredentialsDelegate = nullptr;
+
+    /* The following keypair must correspond to the public key used for generating
+    controllerNOC. It's used by controller to establish CASE sessions with devices */
+    Crypto::P256Keypair * ephemeralKeypair = nullptr;
+
+    /* The following certificates must be in x509 DER format */
+    ByteSpan controllerNOC;
+    ByteSpan controllerICAC;
+    ByteSpan controllerRCAC;
```

Comment:
This is not knowledge that should exist a-priori. The implementation of a Commissioner should be able to provide these just-in-time during NOCSR processing. I am confused why we are adding this to the factory.

Reply:
Yeah I agree. I'll file an issue for this. We'll need a delegate to provide these fields on demand. 

This PR however, is just adding the Controller's Init Params into the Factory Setup Params.